dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1092 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_REDUCE_REDUCE_GPU_IMPL_CUH_
+#define DALI_KERNELS_REDUCE_REDUCE_GPU_IMPL_CUH_
+
+#include <cassert>
+#include <memory>
+#include <utility>
+#include <vector>
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/reduce/reduce_axes_gpu_impl.cuh"
+#include "dali/kernels/reduce/reductions.h"
+#include "dali/kernels/reduce/reduce_setup_utils.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/span.h"
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_view.h"
+#include "dali/core/traits.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+enum class ReductionKind {
+  All,     ///< Reduce contiguous memory
+  Sample,  ///< Reduce samples separately
+  Block,   ///< Reduce contiguous samples separately
```

Comment:
So we assume that samples can be not continuous?

Reply:
Yes.